[css-anchor-position-1] Implement `anchor-scope`

tuankiet65 opened the pull request:
#### 218536cd505174b33e2df2d2031db1bc690ca595
<pre>
[css-anchor-position-1] Implement `anchor-scope`
<a href="https://rdar.apple.com/129789606">rdar://129789606</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=275449">https://bugs.webkit.org/show_bug.cgi?id=275449</a>

Reviewed by Antti Koivisto.

This patch implements the logic for scoping anchor elements/queries using
anchor-scope. This is done by adding a condition to the anchor matching process:
if the anchor element is scoped by an ancestor element, the anchor positioned
element must also belong to the scope (i.e, the anchor and anchor-positioned
element are within the same subtree with the ancestor as the root).

* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scope-basic-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scope-dynamic-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scope-shadow-flat-tree-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/parsing/anchor-scope-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/parsing/anchor-scope-computed.html:
    - Add tests to check the order of identifiers in computed value
      matches the order in the declared value.

* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/parsing/anchor-scope-parsing-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/parsing/anchor-scope-parsing.html:
* Source/WTF/wtf/Forward.h:
    - Forward declare ListHashSet.

* Source/WTF/wtf/ListHashSet.h:
    - Don&apos;t declare a default value for template parameter, since it&apos;s now
      defined in the forward declaration in wtf/Forward.h

* Source/WTF/wtf/text/TextStream.h:
(WTF::operator&lt;&lt;):
    - Implement operator&lt;&lt;(TextStream ts, ) for ListHashSet.

* Source/WebCore/rendering/style/NameScope.h:
    - Use ListHashSet to represent the name list instead of Vector.
      This allows O(1) lookup, and ListHashSet also preserves the
      insertion order, so the computed order could match the
      original declared value.
(WebCore::operator==):
    - Implement custom equality comparison between NameScopes.
      This is needed because ListHashSet doesn&apos;t support operator==.

* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::rareDataChangeRequiresLayout):
    - Trigger layout if the anchor scope is different between RenderStyle.

* Source/WebCore/style/AnchorPositionEvaluator.cpp:
(WebCore::Style::anchorScopeForAnchorElement):
    - Add function to locate ancestor that establishes an anchor scope
      affecting an anchor element.
(WebCore::Style::elementIsDescendantOfElementInComposedTree):
    - Add function to check if an element is a descendant of another element
      in the composed tree.
(WebCore::Style::isAcceptableAnchorElement):
    - Add anchor scope check.

* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertNameScope):
    - Change to accomodate the switch from Vector to ListHashSet.

Canonical link: <a href="https://commits.webkit.org/291349@main">https://commits.webkit.org/291349@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2a0e3297eadf40df9e9d49f3677001370c1788bb

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/92739 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/131/builds/12290 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/138/builds/1929 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/97735 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/43250 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/94789 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/12565 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/20742 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/97735 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/43250 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/95741 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/9508 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/138/builds/1929 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/97735 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/9199 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/138/builds/1929 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/42578 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/85449 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/79509 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/138/builds/1929 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/99756 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/91405 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/19791 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/14560 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/80005 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/121/builds/20042 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/138/builds/1929 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/79307 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19656 "Built successfully and passed tests") | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/138/builds/1929 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/12805 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/19775 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/24951 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/114053 "Built successfully") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/19462 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/32952 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/22922 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/21203 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

anttijk commented on Source/WebCore/style/AnchorPositionEvaluator.cpp:
```diff
@@ -741,9 +741,59 @@ static bool firstChildPrecedesSecondChild(const RenderObject* firstChild, const
+    ASSERT(!anchorNames.isEmpty());
+
+    // Traverse up the element tree through each ancestor.
+    for (CheckedPtr currentAncestor = anchorElement.parentElement(); currentAncestor; currentAncestor = currentAncestor->parentElement()) {
```
I think this should be using `parentElementInComposedTree()`. The spec text says

> Like most operations in CSS besides selector matching, features in this specification operate over the [flattened element tree](https://drafts.csswg.org/css-scoping-1/#flat-tree).

(webkit uses "composed tree" for "flat tree")

anttijk commented on Source/WebCore/style/AnchorPositionEvaluator.cpp:
```diff
@@ -741,9 +741,59 @@ static bool firstChildPrecedesSecondChild(const RenderObject* firstChild, const
+
+    if (auto anchorScopeElement = anchorScopeForAnchorElement(*anchorRenderer.element(), anchorRenderer.style().anchorNames())) {
+        if (!anchorPositionedElement->isDescendantOf(*anchorScopeElement))
+            return false;
```
Similarly this should probably operate in the flat tree. I don't think there `isDescendantInComposedTree` helper currently but you can just iterate with parentElementInComposedTree.

anttijk commented on LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scope-shadow-flat-tree-expected.txt:
```diff
@@ -1,3 +1,3 @@
 
-FAIL anchor-scope scopes to the flat tree assert_equals: expected "10px" but got "20px"
+PASS anchor-scope scopes to the flat tree
```
looks like there was a test for this

tuankiet65 committed:
[css-anchor-position-1] Implement `anchor-scope`
rdar://129789606
https://bugs.webkit.org/show_bug.cgi?id=275449

Reviewed by Antti Koivisto.

This patch implements the logic for scoping anchor elements/queries using
anchor-scope. This is done by adding a condition to the anchor matching process:
if the anchor element is scoped by an ancestor element, the anchor positioned
element must also belong to the scope (i.e, the anchor and anchor-positioned
element are within the same subtree with the ancestor as the root).

* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scope-basic-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scope-dynamic-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scope-shadow-flat-tree-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/parsing/anchor-scope-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/parsing/anchor-scope-computed.html:
    - Add tests to check the order of identifiers in computed value
      matches the order in the declared value.

* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/parsing/anchor-scope-parsing-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/parsing/anchor-scope-parsing.html:
* Source/WTF/wtf/Forward.h:
    - Forward declare ListHashSet.

* Source/WTF/wtf/ListHashSet.h:
    - Don't declare a default value for template parameter, since it's now
      defined in the forward declaration in wtf/Forward.h

* Source/WTF/wtf/text/TextStream.h:
(WTF::operator<<):
    - Implement operator<<(TextStream ts, ) for ListHashSet.

* Source/WebCore/rendering/style/NameScope.h:
    - Use ListHashSet to represent the name list instead of Vector.
      This allows O(1) lookup, and ListHashSet also preserves the
      insertion order, so the computed order could match the
      original declared value.
(WebCore::operator==):
    - Implement custom equality comparison between NameScopes.
      This is needed because ListHashSet doesn't support operator==.

* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::rareDataChangeRequiresLayout):
    - Trigger layout if the anchor scope is different between RenderStyle.

* Source/WebCore/style/AnchorPositionEvaluator.cpp:
(WebCore::Style::anchorScopeForAnchorElement):
    - Add function to locate ancestor that establishes an anchor scope
      affecting an anchor element.
(WebCore::Style::elementIsDescendantOfElementInComposedTree):
    - Add function to check if an element is a descendant of another element
      in the composed tree.
(WebCore::Style::isAcceptableAnchorElement):
    - Add anchor scope check.

* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertNameScope):
    - Change to accomodate the switch from Vector to ListHashSet.

Canonical link: https://commits.webkit.org/291349@main
